FSW/src/util/hkUtil.cpp

```diff
@@ -0,0 +1,184 @@
+ *  Turns heater on when optics temp crosses below HEATER_TEMP_LOW
+ *  Turns heater off when optics temp crosses above HEATER_TEMP_HIGH
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void setHeater()
+{
+    if (latestHkSample.opticsTemp >= HEATER_TEMP_HIGH)
+    {
+        digitalWrite(PIN_HEAT, LOW); // turn heater off
+    }
+    else if (latestHkSample.opticsTemp <= HEATER_TEMP_LOW)
+    {
+        digitalWrite(PIN_HEAT, HIGH); // turn heater on
+    }
+}
+
+/* - - - - - - sampleHousekeepingData - - - - - - *
+ * Usage:
+ *  samples temperature and power data and stores it in hkData,
+ *  checks for out-of-bounds values and logs faults if found
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void sampleHousekeepingData()
+{
+    // read thermistor voltage values
+    float opticsTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_OPTICS_THERM);
+    float analogTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_ANALOG_THERM);
+    float digitalTempVoltage = TEENSY_VOLTAGE_RES*analogRead(PIN_DIGITAL_THERM);
+    
+    // update housekeeping data
+    latestHkSample.opticsTemp = voltageToTemp(opticsTempVoltage); 
+    latestHkSample.analogTemp = voltageToTemp(analogTempVoltage);
+    latestHkSample.digitalTemp = voltageToTemp(digitalTempVoltage);
+    latestHkSample.analogCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_AREG_CURR);
+    latestHkSample.digitalCurrent = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_CURR);   
+    latestHkSample.digitalRegPG = TEENSY_VOLTAGE_RES*(float)analogRead(PIN_DREG_PG);
+    hkData[hkIndex] = latestHkSample; // update data array
+
+    // Check if values are in acceptable ranges
+    // TODO: Log a fault for every case
+    if (latestHkSample.opticsTemp > OPTICS_TEMP_MAX_SAFE) { } // optics temp too high
+    else if (latestHkSample.opticsTemp < OPTICS_TEMP_MIN_SAFE) { } // optics temp too low
+
+    if (latestHkSample.analogTemp > BOARD_TEMP_MAX_SAFE) { } // analog board temp too high
+    else if (latestHkSample.analogTemp < BOARD_TEMP_MIN_SAFE) { } // analog board temp too low
+
+    if (latestHkSample.digitalTemp > BOARD_TEMP_MAX_SAFE) { } // digital board temp too high
+    else if (latestHkSample.digitalTemp < BOARD_TEMP_MIN_SAFE) { } // digital board temp too low
+
+    if (latestHkSample.digitalRegPG > PG_VOLTAGE_MAX_EXPECTED) { } // digital regulator PG signal too high
+    else if (latestHkSample.digitalRegPG < PG_VOLTAGE_MIN_EXPECTED) { } // digital regulator PG signal too low
+
+}
+
+/* - - - - - - voltageToTemp - - - - - - *
+ * Usage:
+ *  Interpolates and returns the corresponding temperature for a given thermistor voltage reading
+ *  based on the data in thermLookup. 
+ *  Will extrapolate beyond max and min entries in thermLookup if needed
+ * 
+ * Inputs:
+ *  voltage - thermistor voltage
+ *  
+ * Outputs:
+ *  temperature - celsius, the temperature corresponding to thermistor voltage reading
+ */
+float voltageToTemp(float voltage)
```

Comment:
The table you created, using the MATLAB script (which looks good) is only valid for the analog and digital thermistors. As per these equations: 
![image](https://user-images.githubusercontent.com/25249287/108008020-93d9ee80-6fbc-11eb-9188-9feb8b47f5e1.png)

Recall that the optics thermistor uses a different set of equations: which is recreated via this transfer table: 
![image](https://user-images.githubusercontent.com/25249287/108008065-ad7b3600-6fbc-11eb-8472-d621d58a9a4b.png)

Reply:
Thanks, seems like I missed this. Will implement.